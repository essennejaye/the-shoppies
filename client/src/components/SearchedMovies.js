import React, { useState, useEffect } from "react";
import {
  Jumbotron,
  Container,
  Button,
  Form,
  Card,
  Col,
  ListGroupItem,
  ListGroup,
} from "react-bootstrap";
import { useMutation } from "@apollo/react-hooks";
import { SAVE_MOVIE } from "../utils/mutations";
import { saveMovieIds, getSavedMovieIds } from "../utils/localStorage";
import SavedMovies from "./SavedMovies";

const SearchedMovies = () => {
  // create states for holding returned fetch results, search input field and saved movies
  const [searchedMovies, setSearchedMovies] = useState([]);
  const [searchInput, setSearchInput] = useState("");
  const [savedMovieIds, setSavedMovieIds] = useState(getSavedMovieIds());

  const [saveMovie, { error }] = useMutation(SAVE_MOVIE);

  // save movieid to local storage for persistence
  useEffect(() => {
    saveMovieIds(savedMovieIds);
  }, [savedMovieIds]);

  const handleFormSubmit = async (event) => {
    event.preventDefault();

    if (!searchInput) {
      alert("You must enter a search term!");
      return false;
    }
    const key = process.env.REACT_APP_API_KEY;

    try {
      const response = await fetch(
        `http://www.omdbapi.com/?s=${searchInput}&apikey=${key}`
      );
      if (!response.ok) {
        throw new Error("Oops, something went wrong");
      }

      const items = await response.json();
      
      if (!items || !items['Search']) {
        alert("No results for that search term");
        setSearchInput('');
        return false;
      }

      const movies = items.Search;
      const movieData = movies.map((movie) => ({
        movieID: movie.imdbID,
        title: movie.Title,
        year: movie.Year,
      }));
      setSearchedMovies(movieData);
    } catch (err) {
      console.error(err);
    }
  };

  const handleSaveMovie = async (movieID) => {
    if (savedMovieIds.length >= 5) {
      alert("You can only have 5 nominees!");
      return;
    }

    const movieToSave = searchedMovies.find(
      (movie) => movie.movieID === movieID
    );

    try {
      await saveMovie({
        variables: {
          title: movieToSave.title,
          year: movieToSave.year,
          movieID: movieToSave.movieID,
        },
      });
    } catch (err) {
      console.error(err);
    }
    setSavedMovieIds([...savedMovieIds, movieToSave.movieID]);
  };

  // const newLocal = searchedMovies.length || savedMovieIds.length;
  return (
    <>
      <Jumbotron fluid className="text-light bg-dark">
        <Container>
          <h1>The Shoppies</h1>
          <Form onSubmit={handleFormSubmit}>
            <Form.Row>
              <Col xs={12} md={8}>
                <Form.Control
                  name="searchInput"
                  value={searchInput}
                  onChange={(e) => setSearchInput(e.target.value)}
                  type="text"
                  size="lg"
                  placeholder="Search for a movie"
                />
              </Col>
              <Col xs={12} md={4}>
                <Button type="submit" variant="success" size="lg">
                  Submit Search
                </Button>
              </Col>
            </Form.Row>
          </Form>
        </Container>
      </Jumbotron>

      {(!searchedMovies.length ) ? null :
        <section className="shoppie-container">
          <Container className="result-container">
          <h2>Results for '{searchInput}'</h2>
            <Card>
              <ListGroup variant="flush">
                {searchedMovies.map((movie) => {
                  return (
                    <ListGroupItem key={movie.movieID}>
                      {`Title: ${movie.title} (${movie.year})`}
                      <Button
                        disabled={savedMovieIds?.some(
                          (savedID) => savedID === movie.movieID
                        )}
                        className="btn-block, btn-info"
                        onClick={() => {
                          handleSaveMovie(movie.movieID);
                        }}
                        variant="primary"
                      >
                        Nominate
                      </Button>
                    </ListGroupItem>
                  );
                })}
              </ListGroup>
            </Card>
          </Container>
          
          {/* pass setter function as prop to child so state can be updated from child */}
          <SavedMovies
            newMovieIds={savedMovieIds}
            setSavedMovieIds={setSavedMovieIds}
          />
        </section>
      }
      {/* need better error handling */}
      {error && <h2>Oops!</h2>}
    </>
  );
};

export default SearchedMovies;
