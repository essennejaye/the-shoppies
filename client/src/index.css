body {
  font-family: "Rubik", sans-serif;
}

h1,
h2,
h3,
h4,
h5,
h6 {
  font-family: "Rubik Mono One", sans-serif;
}

.small {
  font-weight: 300;
  font-style: italic;
}
.shoppie-container {
  display: flex;
  justify-content: space-between;
}
.list-group-item {
  display: flex;
  justify-content: space-between;
}
.btn-success {
  margin: 0px 10px;
}
.modal-background {
  position: fixed;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.5);
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 1;
  opacity: 0;
}

.modal-background.show {
  opacity: 1;
}

.modal-container {
  position: relative;
  width: 500px;
  background-color: #fefefe;
  margin: auto;
  padding: 0;
  border: 1px solid #888;
  box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2),0 6px 20px 0 rgba(0,0,0,0.19);
  animation: animatetop 0.4s;
}
/* Add animation */
@keyframes animatetop {
  from {top:-300px; opacity:0}
  to {top:0; opacity:1}
}

.modal-header,
.modal-foot {
  padding: 2px 16px;
  background-color: #5cb85c;
  color: white;
}
.modal-title {
  margin: 13px 0 8px 0;
  align-self: center;
  font-size: 28px;
}
.modal-body {
  padding: 12px 16px;
  font-size: 20px;
  font-weight: 700;
}
.modal-body p {
  white-space: pre-wrap;
}
.btn-close {
  color: white;
  float: right;
  font-weight: bold;
  font-size: 28px;
}
.btn-close-secondary {
  color: white;
  align-self: flex-start;
  font-weight: bold;
}
.btn-close-secondary,
.btn-close:hover,
.bth-close:focus {
  color:000;
  text-decoration: none;
  cursor: pointer;
}