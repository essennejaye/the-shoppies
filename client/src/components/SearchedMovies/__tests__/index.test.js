import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import SearchedMovies from '..';
import { MockedProvider } from '@apollo/client/testing';

beforeAll(() => jest.spyOn(window, 'fetch'))

// const mocks = [];
describe('SearchedMovies component', () => {
  it('renders without error', () => {
    render(
      <MockedProvider>
        <SearchedMovies />
      </MockedProvider>
    );
    expect(screen.getByPlaceholderText('Search for a movie or tv show by name')).toBeInTheDocument();
  });

  it("matches snapshot DOM node structure", () => {
    const { asFragment } = render(
      <MockedProvider>
        <SearchedMovies />
      </MockedProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders modal error alert for empty input field', () => {
    const searchInput = '';
    render(
      <MockedProvider>
        <SearchedMovies searchInput={searchInput} />
      </MockedProvider>
    );
    userEvent.click(screen.getByRole('button', { name: /submit search/i }))
    expect(screen.getByText('Please enter a search term!')).toBeInTheDocument()
    screen.debug();
  });

  // it('renders modal error alert for invalid input field', async () => {
  //   const searchInput = 'kgjihor';
  //   render(
  //     <MockedProvider>
  //       <SearchedMovies searchInput={searchInput} />
  //     </MockedProvider>
  //   )
  //   window.fetch.mockResolvedValueOnce({
  //     ok: true,
  //     json: async () => ({ success: true }),
  //   });

  //   userEvent.click(screen.getByRole('button', { name: /submit search/i }))
  //   expect(window.fetch).toHaveBeenLastCalledWith(
  //     '/'
      
  //   )
  //   // expect(screen.getByText('Please enter a search term!')).toBeInTheDocument()
  // });
})


