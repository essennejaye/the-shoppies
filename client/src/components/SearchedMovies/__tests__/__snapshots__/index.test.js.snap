// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchedMovies component matches snapshot DOM node structure 1`] = `
<DocumentFragment>
  <section
    class="bg-image"
  >
    <h1
      class="title"
    >
      The Shoppies
    </h1>
    <img
      alt="movie themed clipart"
      class="hero-img"
      src="../images1/movies.png"
    />
  </section>
  <section
    class="form-container"
  >
    <form
      class="form-submit"
    >
      <div
        class="form-label"
      >
        <label
          for="searchInput"
        >
          <h6>
            Search
          </h6>
        </label>
      </div>
      <div
        class="input-form"
      >
        <input
          name="searchInput"
          placeholder="Search for a movie or tv show by name"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-btns"
      >
        <button
          class="form-searchBtn btn"
          type="submit"
          variant="success"
        >
          Submit Search
        </button>
        <button
          class="form-clearBtn btn"
          type="button"
          variant="info"
        >
          Clear Search
        </button>
      </div>
    </form>
  </section>
  <section
    class="shoppie-container row"
  />
</DocumentFragment>
`;
